---
title: "Toyfish - a quick tutorial with an example dataset"
output: github_document
---


## Dataset overview

![](https://media.fisheries.noaa.gov/dam-migration/cryptic-pairs-the-vermilion-and-sunset-rockfishes-noaa-nwfsc.png)

To help users quickly learn and test loco-pipe on their own, we provide a heavily subsetted lcWGS dataset with which the pipeline can be run in a few minutes. This dataset that we call "toyfish" is composed of 30 rockfish individuals from a pair of sister taxa: sunset rockfish (Sebastes crocotulus) and vermilion rockfish (Sebastes miniatus). Each species has 15 specimens. Within vermilion, there are three populations that are named creatively as: vermilion_1, vermilion_2 and vermilion_3; each population has 5 specimens. 

As shown in the picture above, these two species are very similar in their phenotypes and even experts cannot accurately to tell them apart. It is even harder to distinguish one vermilion population from another based only on visual cues. So hopefully, in addition to teaching users about loco-pipe, this tutorial will be a good demonstration of the power of lcWGS in characterizing population structure in natural populations.

Rockfishes have 24 chromosomes; however, with the aim of having a small test dataset, we selected a 3Mbp region from chromosomes 15 and 16 (from 17Mbp to 20Mbp), and further subset their read depth to 40% of the original data. These bam files are stored in `toyfish/bams`. The subsetted reference genome sequence and its index file are stored in `toyfish/reference`


## To run the pipeline

1. 

2. 

3. 

```{bash, eval = FALSE}
## First, move to the folder that includes the exploratory_analyses.smk.
PIPEDIR=/global/scratch/users/nicolas931010/loco-pipe
BASEDIR=/global/scratch/users/nicolas931010/loco-pipe/toyfish
CONDA=/global/scratch/users/nicolas931010/conda-envs

## Then, activate the Loco-pipe environment
conda activate loco-pipe

snakemake \
  --use-conda \
  --conda-frontend mamba \
  --conda-prefix $CONDA \
  --directory $BASEDIR \
  --scheduler greedy \
  --cores 1 \
  --rerun-triggers mtime \
  --snakefile $PIPEDIR/workflow/pipelines/loco-pipe.smk -p -n

## Finally, adjust the flags below as needed and run the pipeline! 
## The ending -n flag means it is a dry run. If the dry run goes through, remove "-n" and rerun the pipe-line.
snakemake \
  --conda-frontend mamba \
  --conda-prefix $CONDA \
  --directory $BASEDIR \
  --profile $PIPEDIR/workflow/profiles/slurm_savio4_htc \
  --scheduler greedy \
  --default-resources mem_mb=None disk_mb=None \
  --rerun-triggers mtime \
  --snakefile $PIPEDIR/workflow/pipelines/loco-pipe.smk -p -n

## flowchart
mkdir -p $BASEDIR/figures/flowchart
snakemake -n --forceall --rulegraph \
--directory $BASEDIR \
--snakefile $PIPEDIR/workflow/pipelines/loco-pipe.smk  | \
dot -Tpdf > $BASEDIR/figures/flowchart/toyfish.pdf
```


#### To visualize your results.
We provide template codes to make three different visualizations: theta plots, population admixture, and pca plots of all specimens. As automatic and convenient as the template could be, you still need to change some elements manually (e.g. meta_data table pathway, base direction, etc.)
1. Theta plots
```{r, eval = TRUE, fig.width=5, fig.height=3}
# load in some libraries
library(tidyverse)
library(cowplot)
library(janitor)
library(MetBrewer)
# global pathway to the metadata tsv inside your "docs" folder


metadata_pathway<- "toyfish/docs/metadata.tsv"
species_table <- read_tsv(metadata_pathway)%>%mutate(population=pca_cluster) # the "population" is the grouping factor you put down for the config file
basedir<- "toyfish"
chrs <- read_tsv("toyfish/docs/chr_table.tsv", col_names = FALSE) %>% pull(X1)


# Thanks to the way Loco-pipe names files, we could access to the .pestPG file for each sub-group. Change the count() based on your grouping factor.
species_grouping <- species_table %>% mutate(pop_level=common_name) %>%
  count(pop_level) %>% 
  dplyr::select(-n)
average_theta_table <- NULL
for (i in seq_len(nrow(species_grouping))){
  group <- species_grouping$pop_level[i]
  paths <- str_c(basedir, "/angsd/get_theta/", group, ".", chrs,".average_thetas.pestPG")
  average_theta_table_tmp <- read_tsv(paths) %>%
    clean_names() %>%
    summarise(t_w=sum(t_w), t_p=sum(t_p), n_sites=sum(n_sites)) %>%
    transmute(species_group=group, `Watterson's theta`=t_w/n_sites, pi=t_p/n_sites)
  average_theta_table <- bind_rows(average_theta_table, average_theta_table_tmp)
}

# Plot the thetas.
average_theta_plot<-average_theta_table %>%
  pivot_longer(cols = 2:3) %>%
  arrange(species_group) %>%
  mutate(species_group=as_factor(species_group)) %>%
  ggplot(aes(x=fct_rev(species_group), y=value, fill=species_group)) +
  geom_col(color="black") +
  scale_fill_manual(values = met.brewer("Hiroshige", 9, "discrete")) +
  facet_wrap(~name, nrow=1, scales = "free") +
  coord_flip() +
  theme_cowplot() +
  theme(legend.position = "none")

# Print the plots
average_theta_plot
```

2. Admixture
```{r, eval = TRUE, fig.width=6, fig.height=4}
# import some needed library
library(tidyverse)
library(pals)
library(cowplot)
library(sf)
library(rnaturalearth)
library(scatterpie)

# read in data
metadata_pathway<- "/global/scratch/users/zzhou32/loco-pipe/toyfish/docs/metadata.tsv"
mtdt_table <-read_tsv(metadata_pathway)
col_names<-c("sample_name", "common_name", "pca_cluster") # These need to match up the col names in the matadata_table
basedir<- "/global/scratch/users/zzhou32/loco-pipe/toyfish"

# A function that coins multiple q.matrix files together.
read_ohana_q <- function(basedir, k_range, anc_order, md_pathway, col_name_vector){
  sample_table <- read_tsv(md_pathway)
  for (k in k_range) {
    genome_admix_k <-  read_tsv(paste0(basedir,"/ohana/global/combined.subsetted.k", k,".q.matrix"), skip=1, col_names = FALSE) %>%
      as.matrix() %>%
      .[, anc_order[[k - min(k_range) + 1]]] %>%
      as_tibble() %>%
      bind_cols(sample_table%>%dplyr::select(col_name_vector)) %>%
      pivot_longer(cols = 1:k, names_to = "anc", values_to = "p") %>%
      mutate(k = k)
    if (k == min(k_range)) {
      genome_admix <- genome_admix_k
    } else {
      genome_admix <- bind_rows(genome_admix, genome_admix_k)
    }
  }
  return(genome_admix)
}
# A function that does the plotting
plot_admix <- function(genome_admix, order_by_k, color_palette ){
  sample_order <- genome_admix %>%
    filter(k==order_by_k) %>%
    group_by(k,population, anc) %>%
    summarise(mean_p=mean(p)) %>%
    slice_max(order_by = mean_p) %>%
    ungroup() %>%
    semi_join(genome_admix, ., by=c("population", "anc", "k")) %>%
    arrange(population, -p) %>%
    .$sample_name
  genome_admix_for_plot <<- genome_admix %>%
    mutate(sample_id=fct_relevel(sample_name, sample_order)) 
  genome_admix_for_plot %>%
    filter(p>0.005) %>%
    ggplot(aes(x = sample_name, y = p, fill = anc, color = anc)) +
    geom_col(position="fill") +
    scale_fill_manual(values = color_palette, drop=FALSE) +
    scale_color_manual(values = color_palette, drop=FALSE) +
    facet_grid((k) ~ population, scales = "free_x", space = "free_x", switch="y") +
    theme_cowplot() +
    theme(panel.spacing = unit(0.1, "lines"),
          axis.line=element_blank(),
          axis.title=element_blank(),
          axis.text=element_blank(),
          axis.ticks=element_blank(),
          legend.position="none",
          text = element_text(size=10),
          strip.text.x = element_text(size = 10, angle = 0),
          strip.text.y = element_text(size = 10, angle = 180), 
          plot.margin = unit(c(0, 0, 0, 0), "cm"))
}

# Plot the admixture

anc_order_ohana <- lapply(2:7, function(x){1:x})

my_colors <- cols25(n=7) %>% unname()
combined_q <- read_ohana_q(basedir, 
                           2:7, 
                           anc_order_ohana,
                           metadata_pathway,
                           col_names)

ohana_admixture_plot <- combined_q%>%mutate(population = pca_cluster )%>%
  mutate(anc = as_factor(anc)) %>% plot_admix(7, my_colors)

ohana_admixture_plot
```


3. PCA with ANGSD
```{r, eval = TRUE, fig.width=6, fig.height=4}
# read in data
basedir <- "/global/scratch/users/zzhou32/loco-pipe/toyfish"
tsv_pathway<-"/global/scratch/users/zzhou32/loco-pipe/toyfish/docs/metadata.tsv"
mtdt_table <-read_tsv(tsv_pathway)

# build covariance matrix and plot
cov_mat <- read_tsv(str_c(basedir,"/pcangsd/global/combined.subsetted.cov"), col_names = F)
split_strings <- strsplit(cov_mat$X1, " ")
matrix_of_floats <- matrix(as.numeric(unlist(split_strings)), ncol = 30, byrow = TRUE)

eigen_decomposition <-function(cov_matrix, index_exclude=vector()){
 index_include <- setdiff(seq_len(nrow(cov_matrix)), index_exclude)
  m <- as.matrix(cov_matrix)
  m[is.na(m)]<- median(m, na.rm = T)
  m<-m[index_include, index_include] 
  e <- eigen(m)
  return(print(m))
} 

e_vectors<- eigen(matrix_of_floats)$vectors%>%
    as_tibble() %>%
    select(1:10) %>%
    rename_at(vars(str_c("V", 1:10)), ~ str_c("PC", 1:10))%>% bind_cols(mtdt_table,.) %>% mutate(common_name=as_factor(common_name))
set.seed(42)
pca_angsd_plot<-ggplot(e_vectors, aes(x=PC1,y=PC2))+ geom_point(aes(color=common_name))

pca_angsd_plot
```
